Sprint 12

smoozmy opened the pull request.

smoozmy committed:
Добавлены структуры для декодирования JSON

smoozmy committed:
Добавлена загрузка фото в ленту

smoozmy committed:
Подгрузка изображений через заглушку

smoozmy committed:
Изменение состояния лайков

smoozmy committed:
updateLike перенесен на главный поток

smoozmy committed:
Реализован делегат для обновления состояния ячейки после изменения состояния лайка

smoozmy committed:
Добавлена заглушка при загрузке полноразмерного изображения

smoozmy committed:
Кэширование + сброс зума

smoozmy committed:
Реализован логаут

smoozmy committed:
Добавлены анимации при загрузке из сети

smoozmy committed:
Поправил иконки + добавил алерт для логаута

smoozmy committed:
Фикс загрузки изображения

Хотел кэшировать открываемые в оригинале изображения, но тогда при повторном заходе в ячейку не подгружается картинка

Пока что временно вылечил таким способом

Helvetii951 commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,94 @@
+
+        let nextPage = (lastLoadedPage ?? 0) + 1
+        var urlComponents = URLComponents(string: "https://api.unsplash.com/photos")
+        urlComponents?.queryItems = [
```
URL'ы стоит хранить в файле с константами. Если что-то изменится в бэкэнде, будет легко найти и изменить данные.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,94 @@
+        guard let url = URL(string: "https://api.unsplash.com/photos/\(photoId)/like") else {
+            completion(.failure(NetworkError.invalidRequest))
+            return
+        }
```
Та же история с URL'ом.

Helvetii951 commented on ImageFeed/Auth/ProfileLogoutService.swift:
```diff
@@ -0,0 +1,56 @@
+        ImagesListService.shared.clearImages()
+    }
+    
+    private func goToSplachScreen() {
+        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
+           let window = windowScene.windows.first {
+            let splashViewController = SplashViewController()
+            window.rootViewController = splashViewController
+            window.makeKeyAndVisible()
+        }
+    }
```
Стоит убрать из `ProfileLogoutService` обязанность по навигации. Сервис должен заниматься бизнес-моделью. В архитектуре этого проекта(MVC) навигация - дело контроллера.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -53,15 +71,80 @@ final class ImagesListCell: UITableViewCell {
+        super.prepareForReuse()
+        imageCell.kf.cancelDownloadTask()
+        //        stubImageView.isHidden = false
+        for layer in animationLayers {
```
Неиспользуемый закомментированнй код лучше сразу удалять.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -28,44 +28,106 @@ final class ImagesListViewController: UIViewController {
+        
+        cell.startLoadingAnimation()
+        
+        let lowQualityURL = URL(string: photo.thumbImageURL)
```
Функцию можно сделать приватной. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

smoozmy committed:
Исправлены замечания по ревью

- Переиспользование форматтера
- Лайк в ленте работает
- Убран закомментированный код
- Урлы вынесены в константу

+ Добавлено кэширование при загрузке оригинала изображения в SingleImagesViewController
